rpc/filter_api.go

```diff
@@ -9,23 +9,71 @@ import (
 
 	"math/big"
 
+	"github.com/ethereum/go-ethereum/common"
 	ethtypes "github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/eth/filters"
 	"github.com/ethereum/go-ethereum/rpc"
 )
 
 // PublicFilterAPI is the eth_ prefixed set of APIs in the Web3 JSON-RPC spec.
 type PublicFilterAPI struct {
-	cliCtx context.CLIContext
+	cliCtx  context.CLIContext
+	backend Backend
+	filters map[uint64]*Filter // ID to filter
+	nextID  uint64
 }
 
 // NewPublicEthAPI creates an instance of the public ETH Web3 API.
-func NewPublicFilterAPI(cliCtx context.CLIContext) *PublicFilterAPI {
+func NewPublicFilterAPI(cliCtx context.CLIContext, backend Backend) *PublicFilterAPI {
 	return &PublicFilterAPI{
-		cliCtx: cliCtx,
+		cliCtx:  cliCtx,
+		backend: backend,
+		filters: make(map[uint64]*Filter),
+		nextID:  0,
 	}
 }
 
+// NewFilter instantiates a new filter.
+func (e *PublicFilterAPI) NewFilter(fromBlock, toBlock *big.Int, addresses []common.Address, topics [][]common.Hash) uint64 {
```

Comment:
okay, should I change the parameters to a map then? or should I change it to FilterQuery? because JSON formats everything as a map, but I don't think it'll end up as a map here?

I was under the impression that the rpc server would unmarshal the params into separate variables, but maybe I am just assuming things lol. is there a specific place in the code where the rpc params get parsed?

Reply:
not in Ethermint code, it uses the Geth framework for this. `FilterQuery` would be my suggestion, as it will accept the correct format for sure.

Yes JSON formats everything as a map, but you can also serialize an array in json, and that is how the parameters are serialized in the web3 specification. How geth handles this is each parameter is deserialized from the array, and those parameters can be in any format, as this one length array is the map, which is serialized/deserialized from that FilterQuery in geth.

Reply:
ah okay I see, I will update to use FilterQuery then